File: app/src/main/java/com/lucasurbas/counter/data/repository/MemoryCounterRepository.java
package com.lucasurbas.counter.data.repository;

import com.annimon.stream.Stream;
import com.lucasurbas.counter.data.model.Counter;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.subjects.BehaviorSubject;

public class MemoryCounterRepository implements CounterRepository {

    private final List<Counter> counterList;
    private BehaviorSubject<List<Counter>> counterListSubject;

    public MemoryCounterRepository() {
        counterList = new ArrayList<>();
        counterListSubject = BehaviorSubject.create();
    }

    public void initData(){
        counterList.add(Counter.builder()
                .id(1)
                .value(TimeUnit.SECONDS.toMillis(30))
                .build());
        counterList.add(Counter.builder()
                .id(2)
                .value(TimeUnit.MINUTES.toMillis(2))
                .build());
        counterList.add(Counter.builder()
                .id(3)
                .value(TimeUnit.MINUTES.toMillis(5))
                .build());
        counterList.add(Counter.builder()
                .id(4)
                .value(TimeUnit.MINUTES.toMillis(30))
                .build());
        counterListSubject = BehaviorSubject.createDefault(counterList);
    }

    public void initData(List<Counter> list){
        counterList.addAll(list);
        counterListSubject = BehaviorSubject.createDefault(counterList);
    }

    @Override
    public Single<Counter> getCounterWithId(int id) {
        return Observable.fromIterable(counterList)
                .filter(counter -> counter.isIdEqual(id))
                .firstOrError();
    }

    @Override
    public Observable<Counter> getCounterUpdatesWithId(int id) {
        return counterListSubject
                .flatMap(Observable::fromIterable)
                .filter(counter -> counter.isIdEqual(id));
    }

    @Override
    public Single<List<Counter>> getCounters() {
        return Single.fromCallable(() -> counterList);
    }

    @Override
    public Observable<List<Counter>> getCountersUpdates() {
        return counterListSubject;
    }

    @Override
    public Completable updateCounter(final Counter updatedCounter) {
        return Completable.fromRunnable(() -> {
            List<Counter> list = Stream.of(counterList)
                    .map(counter -> counter.isIdEqual(updatedCounter.getId()) ? updatedCounter : counter)
                    .toList();
            counterList.clear();
            counterList.addAll(list);
            counterListSubject.onNext(new ArrayList<>(counterList));
        });
    }
}

File: app/src/main/java/com/lucasurbas/counter/data/repository/MemoryCounterRepository.kt
package com.lucasurbas.counter.data.repository

import com.annimon.stream.Stream
import com.lucasurbas.counter.data.model.Counter
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.subjects.BehaviorSubject
import java.util.*
import java.util.concurrent.TimeUnit

class MemoryCounterRepository : CounterRepository {

    private val counterList: MutableList<Counter>
    private var counterListSubject: BehaviorSubject<List<Counter>>

    init {
        counterList = ArrayList()
        counterListSubject = BehaviorSubject.create()
    }

    fun initData() {
        counterList.add(Counter(id = 1, value = TimeUnit.SECONDS.toMillis(30), color = "#90CCF4"))
        counterList.add(Counter(id = 2, value = TimeUnit.MINUTES.toMillis(2), color = "#F3D250"))
        counterList.add(Counter(id = 3, value = TimeUnit.MINUTES.toMillis(5), color = "#F78888"))
        counterList.add(Counter(id = 4, value = TimeUnit.MINUTES.toMillis(30), color = "#93C95B"))
        counterListSubject = BehaviorSubject.createDefault(counterList)
    }

    fun initData(list: List<Counter>) {
        counterList.addAll(list)
        counterListSubject = BehaviorSubject.createDefault(counterList)
    }

    override fun getCounterWithId(id: Int): Single<Counter> {
        return Observable.fromIterable(counterList)
                .filter { counter -> counter.isIdEqual(id) }
                .firstOrError()
    }

    override fun getCounterUpdatesWithId(id: Int): Observable<Counter> {
        return counterListSubject
                .flatMap { Observable.fromIterable(it) }
                .filter { counter -> counter.isIdEqual(id) }
    }

    override fun getCounters(): Single<List<Counter>> {
        return Single.fromCallable { counterList }
    }

    override fun getCountersUpdates(): Observable<List<Counter>> {
        return counterListSubject
    }

    override fun updateCounter(counter: Counter): Completable {
        return Completable.fromRunnable {
            val list = Stream.of(counterList)
                    .map { if (it.isIdEqual(counter.id)) counter else it }
                    .toList()
            counterList.clear()
            counterList.addAll(list)
            counterListSubject.onNext(ArrayList(counterList))
        }
    }
}
